File: BioGuard/frontend/lib/models/reading.dart
/// BioGuard — Reading Model
/// Mirrors the backend's ReadingOut schema.
class Reading {
  final int id;
  final String deviceId;
  final String readingType; // "temperature" or "lock"
  final double? numericValue;
  final String? statusValue;
  final bool anomalous;
  final DateTime timestamp;

  Reading({
    required this.id,
    required this.deviceId,
    required this.readingType,
    this.numericValue,
    this.statusValue,
    required this.anomalous,
    required this.timestamp,
  });

  factory Reading.fromJson(Map<String, dynamic> json) {
    return Reading(
      id: json['id'] as int,
      deviceId: json['device_id'] as String,
      readingType: json['reading_type'] as String,
      numericValue: (json['numeric_value'] as num?)?.toDouble(),
      statusValue: json['status_value'] as String?,
      anomalous: json['anomalous'] as bool,
      timestamp: DateTime.parse(json['timestamp'] as String),
    );
  }
}

File: BioGuard/backend/app/schemas/reading.py
"""
BioGuard Backend — Reading Schemas
Pydantic models for API responses built from the Reading ORM model.
"""

from typing import Optional

from pydantic import BaseModel, ConfigDict

from app.schemas.types import UTCDateTime


class ReadingOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    device_id: str
    reading_type: str
    numeric_value: Optional[float] = None
    status_value: Optional[str] = None
    anomalous: bool
    timestamp: UTCDateTime


class DeviceStatus(BaseModel):
    device_id: str
    # Set only once a Device row exists for this id (e.g. via
    # scripts/seed_devices.py) — null for an auto-registered device that
    # has never been named.
    name: Optional[str] = None
    location: Optional[str] = None
    temperature: Optional[ReadingOut] = None
    lock: Optional[ReadingOut] = None